examples/merge.c

```diff
@@ -0,0 +1,387 @@
+
+		if (curr[0] != '-') {
+			opts_add_refish(opts, curr);
+		} else if (!strcmp(curr, "--no-commit")) {
+			opts->no_commit = 1;
+		} else if (match_str_arg(repo_path, &args, "--git-dir")) {
+		} else {
+			print_usage();
+		}
+	}
+}
+
+static int resolve_refish(git_annotated_commit **commit, git_repository *repo, const char *refish)
+{
+	git_reference *ref;
+	int err = 0;
+	git_oid oid;
+
+	assert(commit != NULL);
+
+	err = git_reference_dwim(&ref, repo, refish);
+	if (err == GIT_OK) {
+		git_annotated_commit_from_ref(commit, repo, ref);
+		git_reference_free(ref);
+		return 0;
+	}
+
+	err = git_oid_fromstr(&oid, refish);
+	if (err == GIT_OK) {
+		err = git_annotated_commit_lookup(commit, repo, &oid);
+	}
+
+	return err;
+}
+
+static int resolve_heads(git_repository *repo, merge_options *opts) {
+	git_annotated_commit **annotated = calloc(opts->heads_count, sizeof(git_annotated_commit *));
+	size_t annotated_count = 0, i;
+	int err = 0;
+
+	for (i = 0; i < opts->heads_count; i++) {
+		err = resolve_refish((git_annotated_commit **)&annotated[annotated_count++], repo, opts->heads[i]);
+		if (err != 0) {
+			fprintf(stderr, "failed to resolve refish %s (%d)\n", opts->heads[i], err);
+			annotated_count--;
+			continue;
+		}
+	}
+
+	if (annotated_count <= 0) {
+		fprintf(stderr, "unable to parse any refish\n");
+		free(annotated);
+		return -1;
+	}
+
+	opts->annotated = annotated;
+	opts->annotated_count = annotated_count;
+	return 0;
+}
+
+static int perform_fastforward(git_repository *repo, const git_oid *target_oid, int is_unborn)
+{
+	git_checkout_options ff_checkout_options = GIT_CHECKOUT_OPTIONS_INIT;
+	git_reference *target_ref;
+	git_reference *new_target_ref;
+	git_object *target = NULL;
+	int err = 0;
+
+	if (is_unborn) {
+		const char *symbolic_ref;
+		git_reference *head_ref;
+
+		/* HEAD reference is unborn, lookup manually so we don't try to resolve it */
+		err = git_reference_lookup(&head_ref, repo, "HEAD");
+		if (err != 0) {
+			fprintf(stderr, "failed to lookup HEAD ref\n");
+			return -1;
+		}
+
+		/* Grab the reference HEAD should be pointing to */
+		symbolic_ref = git_reference_symbolic_target(head_ref);
+
+		/* Force-create that reference on the target OID */
+		err = git_reference_create(&target_ref, repo, symbolic_ref, target_oid, 1, NULL);
+		if (err != 0) {
+			fprintf(stderr, "failed to create master reference\n");
+			return -1;
+		}
+
+		git_reference_free(head_ref);
+	} else {
+		/* HEAD exists, just lookup and resolve */
+		err = git_repository_head(&target_ref, repo);
+		if (err != 0) {
+			fprintf(stderr, "failed to get HEAD reference\n");
+			return -1;
+		}
+	}
+
+	/* Lookup the target object */
+	err = git_object_lookup(&target, repo, target_oid, GIT_OBJ_COMMIT);
+	if (err != 0) {
+		fprintf(stderr, "failed to lookup OID %s\n", git_oid_tostr_s(target_oid));
+		return -1;
+	}
+
+	/* Force-checkout the result so the workdir is in the expected state */
+	ff_checkout_options.checkout_strategy = GIT_CHECKOUT_FORCE;
```

Comment:
Do we really want a force checkout? In case there's unstaged changes, we certainly wouldn't want to discard those in case of a fast-forward merge. Git definitly makes sure to keep your data in that scenario